File: backend/app/models/hardware.py
"""Modelos Pydantic para dados de sensores de hardware."""

from __future__ import annotations

from typing import List, Optional
from pydantic import BaseModel, Field


class TemperatureSensor(BaseModel):
    """Leitura individual de um sensor de temperatura."""

    label: str = Field(..., description="Rótulo / nome do sensor")
    current: float = Field(..., description="Temperatura atual em °C")
    high: Optional[float] = Field(None, description="Limiar de alerta em °C")
    critical: Optional[float] = Field(None, description="Limiar crítico em °C")


class TemperatureComponent(BaseModel):
    """Grupo de sensores de um componente de hardware."""

    component: str = Field(..., description="Nome do componente (ex: 'CPU', 'GPU')")
    sensors: List[TemperatureSensor] = Field(default_factory=list)


class CpuStats(BaseModel):
    """Estatísticas de utilização da CPU."""

    usage_percent: float = Field(..., description="Uso geral da CPU em %")
    per_core: List[float] = Field(default_factory=list, description="Uso por núcleo em %")
    frequency_mhz: Optional[float] = None
    temperature: Optional[float] = Field(None, description="Temperatura do pacote em °C")
    temperature_note: Optional[str] = Field(
        None,
        description=(
            "Motivo pelo qual a temperatura não está disponível. "
            "'run_as_admin' = o backend precisa de privilégios de administrador "
            "para acessar os sensores de temperatura via LibreHardwareMonitor."
        ),
    )


class MemorySlot(BaseModel):
    """Informações de um slot físico de memória RAM (via dmidecode)."""

    device_locator: str = Field("", description="Localização do slot (ex: DIMM A1)")
    size_gb: float = Field(0.0, description="Capacidade do módulo em GB")
    speed_mhz: Optional[int] = Field(None, description="Velocidade em MHz")
    manufacturer: Optional[str] = Field(None, description="Fabricante do módulo")
    part_number: Optional[str] = Field(None, description="Número de série / part number")
    form_factor: Optional[str] = Field(None, description="Formato (ex: SODIMM, DIMM)")
    memory_type: Optional[str] = Field(None, description="Tipo (DDR4, DDR5, LPDDR5...)") 


class MemoryStats(BaseModel):
    """Estatísticas de memória do sistema."""

    total_gb: float
    used_gb: float
    available_gb: float
    percent: float
    hardware_total_gb: Optional[float] = Field(
        None, description="Total físico real via dmidecode (soma dos slots)"
    )
    slots: List[MemorySlot] = Field(
        default_factory=list, description="Detalhes de cada slot físico"
    )


class DiskStats(BaseModel):
    """Uso de disco para uma única partição/volume lógico."""

    device: str
    mountpoint: str
    total_gb: float
    used_gb: float
    percent: float
    # Informações do disco físico subjacente (melhor esforço)
    model: Optional[str] = Field(None, description="Modelo do disco físico (ex: Samsung SSD 860 EVO)")
    media_type: Optional[str] = Field(None, description="Tipo de mídia: NVMe | SSD | HDD | Unknown")
    physical_size_gb: Optional[float] = Field(None, description="Capacidade real do disco físico em GB")


class GpuStats(BaseModel):
    """Estatísticas de GPU (quando disponível)."""

    name: str
    load_percent: float
    memory_used_mb: float
    memory_total_mb: float
    temperature: Optional[float] = None
    vendor: Optional[str] = Field(None, description="NVIDIA | AMD | Intel | Unknown")
    driver_version: Optional[str] = Field(None, description="Versão do driver")
    vram_gb: Optional[float] = Field(None, description="VRAM total em GB")


class HardwareSnapshot(BaseModel):
    """Snapshot completo de hardware transmitido via WebSocket."""

    timestamp: float = Field(..., description="Unix timestamp da leitura")
    cpu: CpuStats
    memory: MemoryStats
    disks: List[DiskStats] = Field(default_factory=list)
    gpus: List[GpuStats] = Field(default_factory=list)
    temperatures: List[TemperatureComponent] = Field(default_factory=list)

File: frontend/src/types/hardware.ts
// Tipos TypeScript espelhando os modelos Pydantic do backend

export interface TemperatureSensor {
  label: string
  current: number
  high: number | null
  critical: number | null
}

export interface TemperatureComponent {
  component: string
  sensors: TemperatureSensor[]
}

export interface CpuStats {
  usage_percent: number
  per_core: number[]
  frequency_mhz: number | null
  temperature: number | null
  /** Motivo pelo qual a temperatura não está disponível. 'run_as_admin' = precisa de privilégio Admin. */
  temperature_note?: string | null
}

export interface MemorySlot {
  device_locator: string
  size_gb: number
  speed_mhz: number | null
  manufacturer: string | null
  part_number: string | null
  form_factor: string | null
  memory_type: string | null
}

export interface MemoryStats {
  total_gb: number
  used_gb: number
  available_gb: number
  percent: number
  /** Total físico real via dmidecode (soma dos slots instalados). */
  hardware_total_gb: number | null
  /** Detalhes de cada slot físico de memória. */
  slots: MemorySlot[]
}

export interface DiskStats {
  device: string
  mountpoint: string
  total_gb: number
  used_gb: number
  percent: number
  /** Modelo do disco físico (ex: Samsung SSD 870 EVO). */
  model?: string | null
  /** Tipo de mídia: NVMe | SSD | HDD | Unknown. */
  media_type?: string | null
  /** Capacidade real do disco físico em GB (pode diferir do volume lógico). */
  physical_size_gb?: number | null
}

export interface GpuStats {
  name: string
  load_percent: number
  memory_used_mb: number
  memory_total_mb: number
  temperature: number | null
  vendor: string | null
  driver_version: string | null
  vram_gb: number | null
}

export interface HardwareSnapshot {
  timestamp: number
  cpu: CpuStats
  memory: MemoryStats
  disks: DiskStats[]
  gpus: GpuStats[]
  temperatures: TemperatureComponent[]
}

export interface FanReading {
  id: string
  label: string
  rpm: number
  min_rpm: number | null
  /** RPM máximo observado/registrado (histórico do backend). */
  max_rpm: number | null
  percent: number | null
  controllable: boolean
  /** Modo de velocidade ativo: 'auto' | 'quiet' | 'balanced' | 'turbo' */
  speed_mode?: string | null
}

export interface CurvePoint {
  temperature: number
  fan_percent: number
}

export interface FanProfile {
  id: string
  name: string
  description: string | null
  sensor_source: string
  curve: CurvePoint[]
  is_active: boolean
  is_builtin: boolean
}

export interface FanProfileCreate {
  name: string
  description?: string
  sensor_source: string
  curve: CurvePoint[]
}

export interface FanProfileUpdate {
  name?: string
  description?: string
  sensor_source?: string
  curve?: CurvePoint[]
}

export type ConnectionStatus = 'connecting' | 'connected' | 'disconnected' | 'error'
